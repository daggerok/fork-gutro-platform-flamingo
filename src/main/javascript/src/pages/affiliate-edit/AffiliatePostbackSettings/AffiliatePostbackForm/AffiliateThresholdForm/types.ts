import { AffiliateThreshold } from '../../../../../components/Affiliate/types';

export type AffiliateThresholdFormProps = {
  index: number;
  threshold: AffiliateThreshold;
  deleteThreshold: ((index: number) => void);
  hasEditRights: boolean;
}