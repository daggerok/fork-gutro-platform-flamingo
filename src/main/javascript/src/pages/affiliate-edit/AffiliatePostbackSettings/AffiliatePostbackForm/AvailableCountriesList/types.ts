import { Brand } from '~/types';

export type AvailableCountriesListProps = {
  selectedCountries: string[],
  selectedBrands: Brand[],
  onCountryToggled(iso: string): void,
};
