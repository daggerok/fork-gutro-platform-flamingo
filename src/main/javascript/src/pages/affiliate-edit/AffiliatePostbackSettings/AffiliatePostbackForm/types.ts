import { AffiliatePostback } from '~/components/Affiliate/types';

export type AffiliatePostbackFormProps = {
  postbackIndex: number;
  postback: AffiliatePostback;
  deletePostback: ((index: number) => void);
  hasEditRights: boolean;
};
