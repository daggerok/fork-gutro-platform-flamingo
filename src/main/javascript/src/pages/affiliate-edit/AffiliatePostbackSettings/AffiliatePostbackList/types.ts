import { AffiliatePostback } from './../../../../components/Affiliate/types';

export type AffiliatePostbackListProps = {
  postbacks: AffiliatePostback[];
}
