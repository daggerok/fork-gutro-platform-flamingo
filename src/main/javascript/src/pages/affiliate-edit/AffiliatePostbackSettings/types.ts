export interface AffiliatePostbackSettingsProps {
  hasEditRights: boolean;
}