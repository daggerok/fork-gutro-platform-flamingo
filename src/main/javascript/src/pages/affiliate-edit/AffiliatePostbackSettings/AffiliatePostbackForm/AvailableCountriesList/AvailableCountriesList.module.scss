.countryTag {
  margin-left: 6px;
  margin-top: 6px;

  & .countryIso {
    font-weight: bold;
    margin-left: 10px;
  }
}
