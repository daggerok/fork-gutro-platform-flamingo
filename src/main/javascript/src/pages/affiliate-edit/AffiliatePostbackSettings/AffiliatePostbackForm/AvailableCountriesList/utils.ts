import { Brand, Country } from '~/types';

export const getUniqueCountriesFromSelectedBrands = (selectedBrands: Brand[]): Country[] => {
  const availableCountries: Country[] = [];

  (selectedBrands || [])
    .filter(Boolean)
    .forEach(({ countries }) => {
      (countries || []).forEach((country) => {

        if (!availableCountries.some(({ iso }) => country.iso === iso)) {
          availableCountries.push(country);
        }
      });
    });

  return availableCountries;
};
