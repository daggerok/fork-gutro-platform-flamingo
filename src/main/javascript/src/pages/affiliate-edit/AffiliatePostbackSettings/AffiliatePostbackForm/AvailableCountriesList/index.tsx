import { Alert, Button } from 'antd';
import classnames from 'classnames';
import React, { useEffect, useState } from 'react';
import Flag from '~/components/Common/Flag';
import { isCountryUnavailable } from '~/pages/affiliate-edit/AffiliateBaseSettingsForm/utils';
import { Country } from '~/types';

import styles from './AvailableCountriesList.module.scss';
import { AvailableCountriesListProps } from './types';
import { getUniqueCountriesFromSelectedBrands } from './utils';

const AvailableCountriesList: React.FC<AvailableCountriesListProps> = ({
  selectedCountries,
  selectedBrands,
  onCountryToggled,
}: AvailableCountriesListProps) => {
  const [availableCountries, setAvailableCountries] = useState<Array<Country>>([]);

  useEffect(() => {
    const availableCountries =
      getUniqueCountriesFromSelectedBrands(selectedBrands);

    setAvailableCountries(availableCountries);
  }, [selectedBrands]);

  return (
    <>
      {availableCountries
        .filter((country: Country) => country.active)
        .sort((country) =>
          isCountryUnavailable(country, selectedBrands) ? 0 : -1
        )
        .map((country: Country) => {
          return (
            <Button
              className={styles.countryTag}
              key={country.iso}
              size="small"
              type={
                selectedCountries.includes(country.iso) ? 'primary' : 'ghost'
              }
              danger={isCountryUnavailable(country, selectedBrands)}
              onClick={(): void => onCountryToggled(country.iso)}
              data-testid={`country-button-${country.iso}`}
            >
              <Flag
                country={country.iso}
                isSmall
              />
              <span className={classnames(styles.countryIso)}>
                {country.iso}
              </span>
            </Button>
          );
        })}

      {!selectedCountries.length && (
        <Alert
          type="info"
          message="No selected country will enable this postback on all markets."
        />
      )}
    </>
  );
};

export default AvailableCountriesList;
