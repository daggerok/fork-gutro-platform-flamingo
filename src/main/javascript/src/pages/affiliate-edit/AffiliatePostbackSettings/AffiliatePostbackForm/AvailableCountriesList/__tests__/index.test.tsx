import React from 'react';

import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import AvailableCountriesList from '../';
import { AvailableCountriesListProps } from '../types';

jest.mock('../../../../../../components/Common/Flag');

const mockedOnCountryToggled = jest.fn();

const mockedSECountry = {
  iso: "SE",
  active: true,
};

const mockedNOCountry = {
  iso: "NO",
  active: true,
};

const mockedBatmanCountry = {
  iso: "BATMAN",
  active: true,
};

const defaultProps: AvailableCountriesListProps = {
  selectedBrands: [
    {
      brand: "Gutro",
      label: "Gutro",
      countries: [mockedSECountry, mockedNOCountry],
    },
    {
      brand: "GoGo",
      label: "GoGo",
      countries: [mockedSECountry, mockedNOCountry],
    },
  ],
  selectedCountries: [],
  onCountryToggled: mockedOnCountryToggled,
};

describe("<AvailableCountriesList />", () => {
  beforeEach(() => {
    mockedOnCountryToggled.mockReset();
  });

  describe("renders correctly", () => {
    it("by default", () => {
      const { asFragment } = render(
        <AvailableCountriesList {...defaultProps} />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it("with selected countries", () => {
      const props: AvailableCountriesListProps = {
        ...defaultProps,
        selectedCountries: [mockedSECountry.iso],
      };
      const { asFragment } = render(<AvailableCountriesList {...props} />);

      expect(asFragment()).toMatchSnapshot();
    });

    it("with selected country that isn't in all brands", () => {
      const props: AvailableCountriesListProps = {
        ...defaultProps,
        selectedBrands: [
          {
            brand: "Gutro",
            label: "Gutro",
            countries: [mockedSECountry, mockedNOCountry, mockedBatmanCountry],
          },
          {
            brand: "GoGo",
            label: "GoGo",
            countries: [mockedSECountry, mockedNOCountry],
          },
        ],
        selectedCountries: [mockedBatmanCountry.iso],
      };
      const { asFragment } = render(<AvailableCountriesList {...props} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("handles inputs", () => {
    it("when clicking on a country", () => {
      const { getByTestId } = render(
        <AvailableCountriesList {...defaultProps} />
      );

      userEvent.click(getByTestId("country-button-SE"), undefined);
      expect(mockedOnCountryToggled).toHaveBeenCalledWith(mockedSECountry.iso);
    });

    it("when clicking on a country that isn't in all brands", () => {
      const props: AvailableCountriesListProps = {
        ...defaultProps,
        selectedBrands: [
          {
            brand: "Gutro",
            label: "Gutro",
            countries: [mockedSECountry, mockedNOCountry, mockedBatmanCountry],
          },
          {
            brand: "GoGo",
            label: "GoGo",
            countries: [mockedSECountry, mockedNOCountry],
          },
        ],
        selectedCountries: [],
      };
      const { getByTestId } = render(
        <AvailableCountriesList {...props} />
      );

      userEvent.click(getByTestId("country-button-BATMAN"), undefined);
      expect(mockedOnCountryToggled).toHaveBeenCalledWith(mockedBatmanCountry.iso);
    });
  });
});
