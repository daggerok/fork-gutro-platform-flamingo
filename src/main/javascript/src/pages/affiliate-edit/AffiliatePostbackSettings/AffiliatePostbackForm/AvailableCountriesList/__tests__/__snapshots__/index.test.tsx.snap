// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvailableCountriesList /> renders correctly by default 1`] = `
<DocumentFragment>
  <button
    class="ant-btn ant-btn-ghost ant-btn-sm countryTag"
    data-testid="country-button-NO"
    type="button"
  >
    <mock-flag
      country="NO"
      issmall="true"
    />
    <span
      class="countryIso"
    >
      NO
    </span>
  </button>
  <button
    class="ant-btn ant-btn-ghost ant-btn-sm countryTag"
    data-testid="country-button-SE"
    type="button"
  >
    <mock-flag
      country="SE"
      issmall="true"
    />
    <span
      class="countryIso"
    >
      SE
    </span>
  </button>
  <div
    class="ant-alert ant-alert-info ant-alert-no-icon"
    data-show="true"
    role="alert"
  >
    <div
      class="ant-alert-content"
    >
      <div
        class="ant-alert-message"
      >
        No selected country will enable this postback on all markets.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AvailableCountriesList /> renders correctly with selected countries 1`] = `
<DocumentFragment>
  <button
    class="ant-btn ant-btn-ghost ant-btn-sm countryTag"
    data-testid="country-button-NO"
    type="button"
  >
    <mock-flag
      country="NO"
      issmall="true"
    />
    <span
      class="countryIso"
    >
      NO
    </span>
  </button>
  <button
    class="ant-btn ant-btn-primary ant-btn-sm countryTag"
    data-testid="country-button-SE"
    type="button"
  >
    <mock-flag
      country="SE"
      issmall="true"
    />
    <span
      class="countryIso"
    >
      SE
    </span>
  </button>
</DocumentFragment>
`;

exports[`<AvailableCountriesList /> renders correctly with selected country that isn't in all brands 1`] = `
<DocumentFragment>
  <button
    class="ant-btn ant-btn-ghost ant-btn-sm countryTag"
    data-testid="country-button-NO"
    type="button"
  >
    <mock-flag
      country="NO"
      issmall="true"
    />
    <span
      class="countryIso"
    >
      NO
    </span>
  </button>
  <button
    class="ant-btn ant-btn-ghost ant-btn-sm countryTag"
    data-testid="country-button-SE"
    type="button"
  >
    <mock-flag
      country="SE"
      issmall="true"
    />
    <span
      class="countryIso"
    >
      SE
    </span>
  </button>
  <button
    class="ant-btn ant-btn-primary ant-btn-sm ant-btn-dangerous countryTag"
    data-testid="country-button-BATMAN"
    type="button"
  >
    <mock-flag
      country="BATMAN"
      issmall="true"
    />
    <span
      class="countryIso"
    >
      BATMAN
    </span>
  </button>
</DocumentFragment>
`;
